tests/test_basis_set.py

```diff
@@ -419,3 +422,442 @@ def test_quick_step() -> None:
         ],
     }
     # TODO: generate a QuickStep generator in the CP2K plugin
+
+
+# -------------------------------
+# AtomCenteredFunction (combined shells, padding/truncation, inference)
+# -------------------------------
+
+
+def test_acf_infers_n_primitive_and_ell() -> None:
+    """
+    If n_primitive is omitted, it is inferred from exponents length; ℓ inferred from function_type.
+    """
+    bs = AtomCenteredBasisSet()
+    acf = AtomCenteredFunction(
+        function_type='d',
+        exponents=[0.3, 0.2, 0.1],
+        contraction_coefficients=[1.0, 1.0, 1.0],
+    )
+    bs.functional_compositions.append(acf)
+    acf.normalize(None, logger)
+
+    assert acf.n_primitive == 3
+    assert acf.angular_momentum == 2  # d → ℓ=2
+
+
+def test_acf_invalid_length_reset() -> None:
+    """
+    Tests that if n_primitive is set, but coefficient/exponent arrays mismatch length,
+    they are correctly reset to None (due to validation failure).
+    """
```

Comment:
If this generates any logs, you can check them too. Just provide a test logger object.